Homework4.Task1.Kushnir

ArtemKushnir opened the pull request.

ArtemKushnir committed:
Homework4.Task1.Kushnir

toadharvard commented on src/Homeworks/Homework4/Binary_calculator.py:
```diff
@@ -0,0 +1,118 @@
+def change(number1, number2, negative_number2):
+    if len(number1) > len(number2):
+        number2 = [number2[0]] + [0] * (len(number1) - len(number2)) + number2[1:]
+        negative_number2 = (
```
Нормально сбилдь лист, ты тут минимум трижды пересоздаешь лист

toadharvard commented on src/Homeworks/Homework4/Binary_calculator.py:
```diff
@@ -0,0 +1,118 @@
+    else:
+        bit = []
+        number = abs(number)
+        sign = [1, 0]
```
Почему под знак 2 числа в массиве?

toadharvard commented on src/Homeworks/Homework4/Binary_calculator.py:
```diff
@@ -0,0 +1,118 @@
+        return number
+
+
+def transformation_to_additional(number):
```
Эта функция должна переводить в обратный код и прибавлять 1
Почему тут в else ветке кринж какой-то?
А еще сама ветка else не нужна

toadharvard commented on src/Homeworks/Homework4/Binary_calculator.py:
```diff
@@ -0,0 +1,118 @@
+def calculate_sum(number1, number2):
+    result = []
+    count = 0
+    for i in range(len(number1) - 1, -1, -1):
```
эти ифы заменяются на % спокойно

toadharvard commented on src/Homeworks/Homework4/Binary_calculator.py:
```diff
@@ -0,0 +1,118 @@
+def transformation_to_reverse(number):
+    if number[0] == 0:
+        return number
+    else:
```
else не нужен

toadharvard commented on src/Homeworks/Homework4/Binary_calculator.py:
```diff
@@ -0,0 +1,118 @@
+    elif len(number2) > len(number1):
+        number1 = [number1[0]] + [0] * (len(number2) - len(number1)) + number1[1:]
+        return number1, number2, negative_number2
+    else:
```
не нужен

toadharvard commented on src/Homeworks/Homework4/Binary_calculator.py:
```diff
@@ -0,0 +1,118 @@
+            + negative_number2[1:]
+        )
+        return number1, number2, negative_number2
+    elif len(number2) > len(number1):
```
нужен if

toadharvard commented on src/Homeworks/Homework4/Binary_calculator.py:
```diff
@@ -0,0 +1,118 @@
+        )
+        return number1, number2, negative_number2
+    elif len(number2) > len(number1):
+        number1 = [number1[0]] + [0] * (len(number2) - len(number1)) + number1[1:]
```
вот это все в отдельную функциюю с нормальным названием

ArtemKushnir committed:
1. I select one bit for the sign
2. removed unnecessary else
3. Combined reverse and additional code functions
4. Replaced if in the sum function with %
5. divided and rename the change function

toadharvard commented on src/Homeworks/Homework4/Binary_calculator.py:
```diff
@@ -0,0 +1,113 @@
+
+
+def add_digits_to_second_number(number1, number2, negative_number2):
+    for i in range(len(number1) - len(number2)):
```
Во-первых, цикл не нужен. Ты и так  знаешь разницу.
Во-вторых, Если бы ты хранил числа в обратном порядке, то insert не занимал бы O(n) времени, т.к превратился бы в append.

toadharvard commented on src/Homeworks/Homework4/Binary_calculator.py:
```diff
@@ -0,0 +1,113 @@
+    return calculate_sum(number, [0] * (len(number) - 1) + [1])
+
+
+def add_digits_to_second_number(number1, number2, negative_number2):
```
Сигнатура функции кринж. Откуда там негативное число взялось вообще.
Почему ты возвращаешь такую пару, а не другую, теряя информацию о number1

toadharvard replied:
По названию аргументов неясно, что туда будет передано совсем не число int, а массив

toadharvard commented on src/Homeworks/Homework4/Binary_calculator.py:
```diff
@@ -0,0 +1,113 @@
+    return number2, negative_number2
+
+
+def add_digits_to_first_number(number1, number2):
```
Чем это отличается по смыслу от  add_digits_to_first_number?
Ничем, только порядком, лол. Что мешает передать аргументы в другом порядке?

toadharvard commented on src/Homeworks/Homework4/Binary_calculator.py:
```diff
@@ -0,0 +1,113 @@
+    return number1
+
+
+def calculate_final_number(number1, number2):
```
Ну такое, как будто бы смешал библиотеку для обработки двоичных чисел и конкретное задание по сумме и разнице.

Особенно заметно тут
```python
    if len(direct_code1) > len(direct_code2):
        direct_code2, direct_negative_code2 = add_digits_to_second_number(
            direct_code1, direct_code2, direct_negative_code2
        )
 ```

toadharvard replied:
Выглядит откровенно плохо из-за этого

toadharvard commented on src/Homeworks/Homework4/Binary_calculator.py:
```diff
@@ -0,0 +1,113 @@
+def main():
+    print("Enter first number:")
+    first_number = input()
+    if not first_number.lstrip("-").isdigit():
```
Лучше бы функцию обработки ввода вынес, чем вывода на экран.

toadharvard replied:
Кроме того, такая проверка на float невалидна для множества кейсов. Используй исключения и float()

toadharvard commented on src/Homeworks/Homework4/Binary_calculator.py:
```diff
@@ -0,0 +1,113 @@
+    return result[::-1]
+
+
+def output_numbers_and_calculations(
```
Зачем это выносить пользователю?
Количество аргументов такое большое, так как ты их просто из функции в функцию перекладываешь.

toadharvard commented on src/Homeworks/Homework4/Binary_calculator.py:
```diff
@@ -0,0 +1,113 @@
+def calculate_sum(number1, number2):
+    result = []
+    count = 0
+    for i in range(len(number1) - 1, -1, -1):
```
А хранил бы в обратном порядке, не имел бы странных for'oв

toadharvard commented on src/Homeworks/Homework4/Binary_calculator.py:
```diff
@@ -0,0 +1,113 @@
+    for i in range(len(number1) - 1, -1, -1):
+        adding_digit = number1[i] + number2[i] + count
+        result.append(adding_digit % 2)
+        if adding_digit > 1:
```
count = adding_digit // 2

toadharvard commented on src/Homeworks/Homework4/Binary_calculator.py:
```diff
@@ -0,0 +1,113 @@
+    binary_sum = "".join(map(str, calculate_sum(binary_number1, binary_number2)))
+    print(f"{binary_number1_string} + {binary_number2_string} = {binary_sum}")
+    print("The sum in decimal form:")
+    print(f"{number1} + {number2} = {number1 + number2}")
```
Нет, это кринж. Ты должен уметь складывать в бинарной форме и переводить из бинарной формы в десятичную.
Нет никакой гарантии, что ответ, который ты в двоичном виде посчитал ДЕЙСТВИТЕЛЬНО такой в десятичном виде

toadharvard commented on src/Homeworks/Homework4/Binary_calculator.py:
```diff
@@ -0,0 +1,113 @@
+    elif number > 0:
+        sign = [0]
+    else:
+        number = abs(number)
```
это тут лишнее, вынеси за условие это присваивание

ArtemKushnir committed:
storing numbers in reverse order, add transform to decimal function

toadharvard commented on src/Homeworks/Homework4/Binary_calculator.py:
```diff
@@ -0,0 +1,97 @@
+    if number_array[-1] == 1:
+        sign = -1
+        number_array = calculate_sum(number_array, [1] * len(number_array))
+        for i in range(len(number_array)):
```
Второй раз этот цикл используешь, можно было вынести в отдельную функцию reverse_binary

toadharvard commented on src/Homeworks/Homework4/Binary_calculator.py:
```diff
@@ -0,0 +1,97 @@
+    while number:
+        bit.append(number % 2)
+        number //= 2
+    bit.append(0)
```
Зачем?

toadharvard commented on src/Homeworks/Homework4/Binary_calculator.py:
```diff
@@ -0,0 +1,97 @@
+
+def add_digits(number_array, difference_length):
+    number_array = number_array[:-1]
+    sign = number_array[-1]
```
Надо строчки местами поменять, а то это так-то не sign нифига получается, ты ведь берешь последний элемент от списка, у которого строчкой выше убрал последний элемент🗿

toadharvard reviewed (changes requested):
Реально странная логика. Добавляешь нолик, убираешь нолик. Потом в add digits  у тебя местами меняются sign и 0. Короче говоря, это плохо.

ArtemKushnir committed:
add reverse function, add sign with sum and diffrence

SpirinEgor commented:
3 ревью позади, а комментарии остались 🫠
Правила есть правила 🤷‍♂️
